Add server UUID for disks access checks (read/read-by-offset/write/delete) to avoid possible races

azat opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add server UUID for disks access checks (read/read-by-offset/write/delete) to avoid possible races. Plus now disks checks will be done for all disks, and not only S3/Azure.

Otherwise, if you are lucky enough, race condition is possible, and you can get some errors because one server already removed the file while another was trying to read it.

But it was possible only for:
- batch deletes check for "s3" disk
- and all checks for "s3_plain" disk, since this disk does not uses random names

Fixes: #37791

azat committed:
Add server UUID for the S3 disks checks to avoid possible races

Otherwise, if you are lucky enough, race condition is possible, and you
can get some errors because one server already removed the file while
another was trying to read it.

But it was possible only for:
- batch deletes check for "s3" disk
- and all checks for "s3_plain" disk, since this disk does not uses
  random names

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

alexey-milovidov commented:
It should be added to all disk directories, not only s3.

azat committed:
Implement access (read/read-by-offset/write/delete) check for all disks

Previously we had such access (read/write/delete) checks only for S3 and
Azure disks (read/read-by-offset/write/delete), this patch adds check
for all disks.

Also I've added the disk name into IDisk interface, since it is required
for the error message in IDisk::checkAccess(), but I have to add
DiskEncrypted::encrypted_name due DiskEncrypted inherits from
DiskDecorator not from IDisk, and so does not have ability to set disk
name (DiskEncrypted could pass the disk name to the DiskDecorator, but
it is not used anywere, and besides this will require to duplicate the
name for each user of DiskDecorator).

Also from nwo on, skip_access_check will make sense for every disk, plus
now startup() called for each disk (before it was missed for some of
them).

And I've added skip_access_check as as a member for DiskRestartProxy,
since it calls startup() on restart().

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>It should be added to all disk directories, not only s3.

You mean Azure disk, or all object storages, or all disks? What you have in mind?

P.S. for now there is only checks for S3 and Azure.

alexey-milovidov commented:
All disks.

azat committed:
Create disk directory before access check for local disk

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: fix test_disk_types (by adding mkdir for HDFS)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Make disks checks only for clickhouse-server

This will fix clickhouse-disks

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix deadlock in DiskRestartProxy on disk restart

stacktrace:
    contrib/libcxx/src/condition_variable.cpp:47::std::__1::condition_variable::wait(std::__1::unique_lock<std::__1::mutex>&)
    contrib/libcxx/src/shared_mutex.cpp:65::std::__1::shared_timed_mutex::lock_shared()
    src/Disks/DiskRestartProxy.cpp:229::DB::DiskRestartProxy::writeFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, unsigned long, DB::WriteMode, DB::WriteSettings const&)
    src/Disks/IDisk.cpp:0::DB::IDisk::checkAccessImpl(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)
    contrib/libcxx/include/string:1499::DB::IDisk::checkAccess()
    src/Disks/IDisk.cpp:0::DB::IDisk::startup(std::__1::shared_ptr<DB::Context const>, bool)
    src/Disks/DiskRestartProxy.cpp:375::DB::DiskRestartProxy::restart(std::__1::shared_ptr<DB::Context const>)
    contrib/libcxx/include/__memory/shared_ptr.h:701::DB::InterpreterSystemQuery::restartDisk(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&)
    src/Interpreters/InterpreterSystemQuery.cpp:508::DB::InterpreterSystemQuery::execute()
    src/Interpreters/executeQuery.cpp:0::DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)
    src/Interpreters/executeQuery.cpp:1083::DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum)
    src/Server/TCPHandler.cpp:0::DB::TCPHandler::runImpl()
    src/Server/TCPHandler.cpp:1904::DB::TCPHandler::run()
    contrib/poco/Net/src/TCPServerConnection.cpp:57::Poco::Net::TCPServerConnection::start()
    contrib/libcxx/include/__memory/unique_ptr.h:48::Poco::Net::TCPServerDispatcher::run()
    contrib/poco/Foundation/src/ThreadPool.cpp:213::Poco::PooledThread::run()
    contrib/poco/Foundation/include/Poco/SharedPtr.h:156::Poco::ThreadImpl::runnableEntry(void*)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>[Integration tests (tsan) [1/4] — fail: 1, passed: 561, flaky: 0](https://s3.amazonaws.com/clickhouse-test-reports/43143/a0deb9442c7f8287c817c7a4cbe9213b3b6ef5c8/integration_tests__tsan__[1/4].html)

- ` test_distributed_ddl/test.py::test_simple_alters[configs]` query timeout - looks flaky (tested locally - passed)

>[Integration tests (tsan) [2/4] — fail: 1, passed: 529, flaky: 2](https://s3.amazonaws.com/clickhouse-test-reports/43143/a0deb9442c7f8287c817c7a4cbe9213b3b6ef5c8/integration_tests__tsan__[2/4].html)

- `test_create_user_and_login/test.py::test_login_as_dropped_user_xml` - flaky (locally not reproduced)

```
E               Exception: Expected either output '1' or one of errors '['no user with such name', 'not found in user directories']', got output= and error=Received exception from server (version 22.12.1):
E               Code: 192. DB::Exception: Received from 172.16.24.2:9000. DB::Exception: C: User has been dropped. Stack trace:
```

>[Integration tests flaky check (asan) — Failed tests found: error: 0, passed: 0, fail: 2, skipped: 11, flaky: 0](https://s3.amazonaws.com/clickhouse-test-reports/43143/a0deb9442c7f8287c817c7a4cbe9213b3b6ef5c8/integration_tests_flaky_check__asan_.html)

- `test_disk_types`, hm `mkdirs` for hdfs does not always helps

azat committed:
tests: fix hdfs disks (chicken and an egg problem with current cluster.py)

Right now cluster.py first creates HDFS and then creates clickhouse in
one API call, so you cannot interract and add missing mkdirs for the
clickhouse, fix this by using root dir where it is possible.

azat commented:
Failures unrelated:

>AST fuzzer (ubsan) — Logical error: 'Invalid number of rows in Chunk column Object position 3: expected 1, got 0'.

- #40400

>Stateless tests (tsan) [2/5] — fail: 1, passed: 909, skipped: 3

- `01111_create_drop_replicated_db_stress`

```
2022-11-19 21:30:00 [localhost] 2022.11.19 21:29:37.186300 [ 15741 ] {302301ba-6c95-4a5b-9c09-b8a10e85e6b2}  executeQuery: Code: 218. DB::Exception: Table default_repl_2.rmt_1422_11882_17658 is dropped. (TABLE_IS_DROPPED) (version 22.12.1.1) (from [::1]:49612) (comment: 01111_create_drop_replicated_db_stress.sh) (in query: insert into default_repl_2.rmt_1422_11882_17658 values ), Stack trace (when copying this message, always include the lines below):
2022-11-19 21:30:00 Code: 218. DB::Exception: Received from localhost:9000. DB::Exception: Table default_repl_2.rmt_1422_11882_17658 is dropped. (TABLE_IS_DROPPED)
```

>Stress test (msan) — OOM killer (or signal 9) in clickhouse-server.log

```
2022.11.20 00:41:18.990557 [ 328526 ] {} <Error> 86d52eda-305e-4b27-8bcc-5c736281143f::all_0_4_10_41(MutateFromLogEntryTask): virtual bool DB::ReplicatedMergeMutateTaskBase::executeStep(): Code: 47. DB::Exception: Missing columns: 'value4' while processing query: 'value2, value3, value4, value0, key, key, value0, value4, value3, value2', required columns: 'value2' 'value3' 'value4' 'value0' 'key', maybe you meant: ['value2','value3','value0','value0','key']. (UNKNOWN_IDENTIFIER), Stack trace (when copying this message, always include the lines below):
```

And something with MergeTree

>Stress test (ubsan) — Backward compatibility check: Error message in clickhouse-server.log (see bc_check_error_messages.txt)

```
2022.11.20 00:27:29.824241 [ 811379 ] {} <Error> d1c900ac-35ff-4156-a969-e0a310902f9f::all_0_38_9_51(MutateFromLogEntryTask): virtual bool DB::ReplicatedMergeMutateTaskBase::executeStep(): Code: 47. DB::Exception: Missing columns: 'value3' while processing query: 'value5, value3, value0, key, key, value0, value3, value5', required columns: 'value5' 'value3' 'value0' 'key', maybe you meant: ['value5','value0','value0','key']. (UNKNOWN_IDENTIFIER), Stack trace (when copying this message, always include the lines below):
```

azat commented on src/Disks/DiskMemory.cpp:
```diff
@@ -141,6 +141,11 @@ class WriteIndirectBuffer final : public WriteBufferFromFileBase
 
+DiskMemory::DiskMemory(const String & name_)
+    : IDisk(name_)
+    , disk_path("memory://" + name_ + '/')
```
Fine by me, this is is how it was originally named.

kssenii replied:
ok

azat replied:
Done in https://github.com/ClickHouse/ClickHouse/pull/43143/commits/38c009214af06e1037ea728f1b5a87eaef3fe35a

azat commented on src/Disks/DiskMemory.cpp:
```diff
@@ -456,13 +461,20 @@ MetadataStoragePtr DiskMemory::getMetadataStorage()
+        ContextPtr context,
+        const DisksMap & /*map*/) -> DiskPtr
+    {
+        bool skip_access_check = global_skip_access_check || config.getBool(config_prefix + ".skip_access_check", false);
```
The intention was to avoid creating files for this checks with `clickhouse-disks`

kssenii replied:
ok, agree

azat commented on src/Disks/IDisk.cpp:
```diff
@@ -126,4 +129,87 @@ SyncGuardPtr IDisk::getDirectorySyncGuard(const String & /* path */) const
+    {
+        if (isReadOnly())
+        {
+            LOG_DEBUG(&Poco::Logger::get("IDisk"),
```
The idea was to mark it readonly only if it was marked as read-only in the first place, example - web disk.

kssenii replied:
yes, but even explicitly marked as read only disks should also be checked for read access even if this check is not really applicable for web disk?

azat replied:
The thing with this access check that it first need to create a file to check if disk is readable or not, but for read-only it is not possible.

kssenii replied:
yeah, this is why I wrote above "I understand that it is not doable in the general case, but for local disk it is, so it could be noop if not possible."

kssenii replied:
But indeed this is not necessary, I do not force on this

azat replied:
I thought about this, and for now I think it is better to firstly check access and the the brokeness/readonly check, and see how this will work out in production. If this will not suites, then change it.

azat replied:
Actually for local disk, it will be work as you suggesting, since there is a check before creating directory:

```
void DiskLocal::checkAccessImpl(const String & path)
{
    try
    {
        fs::create_directories(disk_path);
    }
    catch (...)
    {
        LOG_ERROR(logger, "Cannot create the directory of disk {} ({}).", name, disk_path);
        readonly = true;
        return;
    }

    IDisk::checkAccessImpl(path);
}
```

But only when directory does not exists, I'm fine to add `canWrite` there, to keep behavior consistent.

azat replied:
Done in https://github.com/ClickHouse/ClickHouse/pull/43143/commits/e6e223adc2c9aa380703f77abed15c48415e515f and https://github.com/ClickHouse/ClickHouse/pull/43143/commits/c393af2812cb56a60a99d48c15c0443d2da98a8b

azat commented on src/Disks/DiskLocal.cpp:
```diff
@@ -650,13 +630,19 @@ bool DiskLocal::setup()
     {
         if (!FS::canRead(disk_path))
-            throw Exception(ErrorCodes::PATH_ACCESS_DENIED, "There is no read access to disk {} ({}).", name, disk_path);
+            return false;
```
Yes, indeed, if canRead fails it should not be used at all. Will rebase the patch.

azat committed:
Change the name pattern for memory disks

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Change DiskLocal::setup() signature (it always return true)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix is_read_only property of local disks

But it wasn't a problem before, since before it was not possible to have
readonly==true.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Check if local disk is readonly before access checks

azat commented:
Test failures looks unrelated:

>Stateless tests (tsan) [2/5] — fail: 1, passed: 909, skipped: 3

- `01926_order_by_desc_limit` - #43408

>Stress test (msan) — OOM killer (or signal 9) in clickhouse-server.log

Looks like slow parts removal, see also https://github.com/ClickHouse/ClickHouse/pull/43277#issuecomment-1316900566

azat commented:
>Stateless tests (tsan) [2/5] — fail: 1, passed: 909, skipped: 3

- #43408

>Stress test (msan) — OOM killer (or signal 9) in clickhouse-server.log

Seems that it is slow parts removal.
